// common/_restClient.js
/*根据请求进行转换
*设备设施专用
*/
function requestBefore() {
    this.responseTool = require('./responseTool');
    this.path = require('path');
    this.fs = require('fs');
    this.requestTypes = {
        get: 'get',
        post: 'post',
        upload: 'upload',
        down: 'down',
        pdownloadByParam: 'pdownloadByParam'
    };
    this.request = new request();
    this.tool = require('./tool');
};

requestBefore.prototype = {
    //登录
    plogin: function (req, res, next, requestType, objParam) {
        var _this = this;
        startSendBefore.call(_this, req, requestType, objParam, null, objParam._c || function (err, result) {
            var clientResult = _this.tool.parseResultForEquie(result || [], objParam.fn, null);
            if (clientResult instanceof Array) clientResult = clientResult[0] || {};
            clientResult.__userPass = objParam.data.user_psw;
            clientResult.__userName = objParam.data.user_name;
            _this.responseTool.loginResponse(req, res, {
                err: err,
                puser: clientResult,
                loginType: objParam.loginType
            });
        });
    },
    //获取数据
    pquery: function (req, res, next, requestType, objParam) {
        var _this = this;
        startSendBefore.call(_this, req, requestType, objParam, null, function (err, result, count) {
            var root = _mapConfig[objParam.fn];
            if (err) {
                console.error('获取' + objParam.fn + '的数据err：' + (err.stack || JSON.stringify(err)));
                return _this.responseTool.sendServerException(res);
            }
            var clientResult = _this.tool.parseResultForEquie(result, objParam.fn, null, count);
            if (!clientResult)
                return _this.responseTool.sendServerException(res, '请检查' + pconst.mapFileName + '文件是否有' + objParam.fn + '请求的配置');
            _this.responseTool.sendSuccess(res, clientResult);
        });
    },
    //普通下载
    pfiledownload: function (req, res, next, requestType, objParam) {
        var _this = this;
        var criteria = {};

        criteria.systemCode = _this.tool.systemCode;
        criteria.key = objParam.data.key;
        //criteria.fileType = objParam.data.fileType;
        objParam.data = criteria;

        startSendBefore.call(_this, req, requestType, objParam, null, function (err, result) {
            result = result || {};
            var bufferArr = result.buffer;
            if (err || !bufferArr) {
                console.error('文件下载失败：' + (err.stack || JSON.stringify(err)));
                return _this.responseTool.sendServerException(res);
            }

            var fileName = result.fileName;
            var buffer = Buffer.concat(bufferArr);
            var path = _this.path.join(_uploadPath, Math.random() + '' + new Date().getTime());
            _this.fs.writeFile(path, buffer, function (err) {
                if (err) {
                    console.error("下载时文件写入 出错" + (err.stack || JSON.stringify(err)));
                    return _this.responseTool.sendServerException(res);
                }
                res.download(path, encodeURI(fileName));
            });
        });
    },
    //根据不同参数来进行附件下载
    pdownloadByParam: function (req, res, next, requestType, objParam) {
        var _this = this;
        startSendBefore.call(_this, req, requestType, objParam, null, function (err, result) {
            result = result || {};
            var bufferArr = result.buffer;
            if (err || !bufferArr) {
                console.error('文件下载失败：' + (err.stack || JSON.stringify(err)));
                return _this.responseTool.sendServerException(res);
            }

            var fileName = result.fileName;
            var buffer = Buffer.concat(bufferArr);
            var path = _this.path.join(_uploadPath, Math.random() + '' + new Date().getTime());
            _this.fs.writeFile(path, buffer, function (err) {
                if (err) {
                    console.error("下载时文件写入 出错" + (err.stack || JSON.stringify(err)));
                    return _this.responseTool.sendServerException(res);
                }
                res.download(path, encodeURI(fileName));
            });
        });
    },
    //数据更新
    pupdate: function (req, res, next, requestType, objParam) {
        var _this = this;
        var systemCode = _this.tool.systemCode;
        var secret = _this.tool.systemSecret;

        var tempParam = createParam.call(_this, req, objParam);
        var criteria = tempParam.criteria || {};
        if (criteria.pwithAttachments == true) { //保存时先进行附件上传操作
            var fileCount = 0;
            var alreadyFileCount = 0;
            var isUploadError = false,
                isValidError = false;
            parseAttachment([criteria]);
            if (fileCount == alreadyFileCount) _sendStart();
        } else _sendStart();

        //循环解析参数，以便深层次的逐一上传文件
        function parseAttachment(dataParamArr) {
            for (var x = 0; x < dataParamArr.length; x++) {
                var currDataParam = dataParamArr[x];
                for (var proName in currDataParam) {
                    if (currDataParam.hasOwnProperty(proName) == false) continue;
                    var proValue = currDataParam[proName];
                    if (proName == pconst.attachments) {
                        var attachments = proValue instanceof Array == true ? proValue : [proValue];
                        fileCount += attachments.length;
                        for (var y = 0; y < attachments.length; y++) {
                            var currAttachment = attachments[y];
                            var oldResource = _this.path.basename(currAttachment.path);
                            var pwIndex = oldResource.indexOf('?');
                            if (pwIndex > -1)
                                oldResource = oldResource.substring(0, pwIndex);
                            currAttachment.path = oldResource;
                            if (currAttachment.isNewFile != false) {
                                var tempName = currAttachment.path;
                                var basePath = _this.path.join(_uploadPath, tempName);
                                if (!_this.fs.existsSync(basePath)) {
                                    console.error('文件' + basePath + '不存在');
                                    isValidError = true;
                                    return _this.responseTool.sendServerException(res);
                                }

                                var time = new Date().getTime();
                                var randomStr = (Math.random() + '').substr(2);
                                var key = time + randomStr;
                                key = key.substr(0, 30);
                                var middle = 30 - key.length;
                                for (var i = 0; i < middle; i++) {
                                    key += '1';
                                }
                                key += (currAttachment.fileName || '') + '.' + (currAttachment.fileSuffix || '');

                                var uploadInfo = {
                                    systemCode: systemCode,
                                    secret: secret,
                                    path: basePath,
                                    fileName: currAttachment.fileName,
                                    fileSuffix: currAttachment.fileSuffix,
                                    key: key,
                                    fileType: currAttachment.fileType
                                };

                                (function (attachment, oldAttachment, oldDataParam, _key) {
                                    var uploadObjParam = {
                                        data: attachment,
                                        fn: pconst.requestType.pupload
                                    };
                                    startSendBefore.call(_this, req, _this.requestTypes.upload, uploadObjParam, null, function (uploadErr, uploadResult) {
                                        if (isValidError || isUploadError) return;

                                        if (uploadErr) {
                                            console.error('文件上传失败：' + (uploadErr.stack || JSON.stringify(uploadErr)));
                                            isUploadError = true;
                                            return _this.responseTool.sendServerException(res);
                                        }
                                        ++alreadyFileCount;

                                        var fileInfos = uploadResult || [];
                                        var newProName = oldAttachment.toPro;
                                        if (oldAttachment.multiFile == true) {
                                            if (oldDataParam[newProName] == null) oldDataParam[newProName] = [];
                                            oldDataParam[newProName].push(_key);
                                            // 20170911 修改请求参数中的值修改的上传中的key值  leo
                                            // oldDataParam[newProName].push((fileInfos[0] || {}).id || '');
                                        } else
                                            oldDataParam[newProName] = _key;

                                        if (fileCount == alreadyFileCount) _sendStart();
                                    });
                                })(uploadInfo, currAttachment, currDataParam, key);
                            } else {
                                ++alreadyFileCount;
                                var ooldPath = psecret.parser(currAttachment.path);
                                var newProName = currAttachment.toPro;
                                if (currAttachment.multiFile == true) {
                                    if (currDataParam[newProName] == null) currDataParam[newProName] = [];
                                    currDataParam[newProName].push(ooldPath);
                                } else currDataParam[newProName] = ooldPath;
                            }
                        }
                        delete currDataParam[proName];
                        continue;
                    }
                    if (proValue instanceof Object == true) {
                        arguments.callee([proValue]);
                        continue;
                    }
                    if (proValue instanceof Array == true) {
                        arguments.callee(proValue);
                        continue;
                    }
                }
            }
        };

        function _sendStart() {
            delete criteria.pwithAttachments;
            var updateObjParam = {
                data: criteria,
                fn: tempParam.fn
            };
            startSendBefore.call(_this, req, _this.requestTypes.post, updateObjParam, null, function (updateErr, updateResult) {
                if (updateErr) {
                    console.error(tempParam.fn + '执行失败：' + (updateErr.stack || JSON.stringify(updateErr)));
                    return _this.responseTool.sendServerException(res);
                }

                _this.responseTool.sendSuccess(res, updateResult);

                // 修改更新后的返回数据 20170922 leo
                // _this.responseTool.sendSuccess(res);
            });
        }
    },
    /*
     *自定义get请求，参数为object，包含属性如下：
     *   url
     *   fn 请求的接口名称
     *   criteria 请求附带的参数
     *   isUpdate   是否是更新操作，数据的新增、修改、删除均为更新，默认false
     *   isParserResult  是否解析结果(只是把content提出来)，默认true
     *   isToDataMap     是否对应datamap，对content内的数据依据dataMap进行转换  默认true
     *   isParserCriteria 是否对请求参数进行处理，默认true
     *   call       请求的回调，两个参数err,result
     */
    sendGet: function (objParam) {
        var _this = this;
        var requestType = objParam.isUpdate == true ? this.requestTypes.post : this.requestTypes.get;
        startSendBefore.call(this, {}, requestType, {
            data: objParam.criteria,
            fn: objParam.fn,
            isParserResult: objParam.isParserResult,
            isParserCriteria: objParam.isParserCriteria
        }, objParam.url, function (err, result) {
            if (err) {
                console.error('获取' + objParam.fn + '的数据err：' + (err.stack || JSON.stringify(err)));
                return typeof objParam.call == 'function' ? objParam.call(err) : '';
            }
            if (objParam.isToDataMap !== false)
                customSendParse.call(_this, objParam.fn, objParam.isUpdate, result, objParam.call);
            else typeof objParam.call == 'function' ? objParam.call(err, result) : '';
        });
    },
    /*自定义post请求，参数同sendGet*/
    sendPost: function (objParam) {
        var _this = this;
        startSendBefore.call(this, {}, this.requestTypes.post, {
            data: objParam.criteria,
            fn: objParam.fn,
            isParserResult: objParam.isParserResult,
            isParserCriteria: objParam.isParserCriteria
        }, objParam.url, function (err, result) {
            if (err) {
                console.error('获取' + objParam.fn + '的数据err：' + (err.stack || JSON.stringify(err)));
                return typeof objParam.call == 'function' ? objParam.call(err) : '';
            }
            if (objParam.isParserResult !== false)
                customSendParse.call(_this, objParam.fn, objParam.isUpdate, result, objParam.call);
            else typeof objParam.call == 'function' ? objParam.call(err, result) : '';
        });
    }
};

module.exports = new requestBefore();

function createParam(req, objParam) {
    var _this = this;
    var param = objParam.data || {};
    var _user = req.session[_this.tool.userSessionName] || {};
    var userId = param.user_id ? param.user_id : _user.userId;
    //var customer_id = _user._customer_id;

    if (userId) {
        if (param instanceof Array == true) {
            for (var i = 0; i < param.length; i++) {
                param[i].user_id = userId;
                //param[i].customer_id = param[i].customer_id ? param[i].customer_id : customer_id;
            }
        } else {
            param.user_id = userId;
            //param.customer_id = param.customer_id ? param.customer_id : customer_id;
        }
    }
    return {
        fn: objParam.fn,
        criteria: param
    };
};

function startSendBefore(req, requestType, objParam, url, call) {
    var tempParam = createParam.call(this, req, objParam);

    // 20170911 leo 修改传入数据中的 -- 转换为空字符串
    if (typeof pconst == 'object')
        tempParam.criteria = JSON.parse(JSON.stringify(tempParam.criteria).replace(new RegExp('"' + pconst.emptyReplaceStr + '"', "g"), '""'));

    var _url;
    switch (requestType) {
        case this.requestTypes.upload:
            var key = encodeURIComponent(tempParam.criteria.key);
            _url = url || (_config.fileServiceUrl + 'systemId=' + this.tool.systemCode + '&secret=' + this.tool.systemSecret + '&key=' + key);
            break;
        case this.requestTypes.down:
            var key = encodeURIComponent(tempParam.criteria.key);
            _url = url || (_config.downFileServiceUrl + '?' + 'systemId=' + this.tool.systemCode + '&key=' + key);
            break;
        default:
            var tempObj = { url: url, fn: tempParam.fn };
            tempObj[this.tool.configServiceParName] = tempParam.criteria[this.tool.configServiceParName];
            _url = this.tool.constructorServiceUrl(tempObj);
            tempParam.criteria[this.tool.configServiceParName] = null;
            delete tempParam.criteria[this.tool.configServiceParName];
            break;
    }

    var par = {
        url: _url,
        criteria: tempParam.criteria,
        isParserResult: objParam.isParserResult,
        isParserCriteria: objParam.isParserCriteria,
        fn: tempParam.fn,
        call: call
    };
    requestType == this.requestTypes.get ? this.request.sendGet(par) :
        requestType == this.requestTypes.upload ? this.request.upload(par) :
            requestType == this.requestTypes.down ? this.request.down(par) :
                requestType == this.requestTypes.pdownloadByParam ? this.request.pdownloadByParam(par) :
                    this.request.sendPost(par);
};

function customSendParse(fn, isUpdate, result, call) {
    if (isUpdate == true) {
        return call(null, result);
    }
    var clientResult = this.tool.parseResultForEquie(result, fn);
    if (!clientResult)
        return call('请检查' + pconst.mapFileName + '文件是否有' + fn + '请求的配置');
    call(null, clientResult);
};

/*对请求进行构造*/
function request() {
    this.send = new send();
    this.javaParamPrefixName = 'jsonString';
    this.middleParamPrefixName = 'pjsonParams';
    this.fs = require('fs');
    this.tool = require('./tool');
};

/*
*普通post请求
*参数objParam    object类型，包括属性如下：
*   url
*   criteria(非必须) object或array类型
*   isParserResult  是否解析结果，默认true
*   isParserCriteria 是否重新构造条件
*   call(非必须)  function类型
*   fn 请求的方法名称，必须
*示例sendPost({
    criteria:{name:'111'},call:function(err,result){},fn:'getUser'
})
*/
request.prototype.sendPost = function (objParam) {
    var requestParam = objParam.criteria;
    this.send.sendPost({
        isParserResult: objParam.isParserResult,
        criteria: requestParam,
        call: objParam.call,
        url: objParam.url
    });
};


/*get请求  参数释义同sendPost*/
request.prototype.sendGet = function (objParam, fnName) {
    var url = objParam.url;
    if (objParam.isParserCriteria !== false) {
        if (objParam.criteria) {
            var criteriaStr = this.constructorCriteriaStr(objParam.criteria);
            url += '?' + this.javaParamPrefixName + '=' + criteriaStr;
        }
    } else {
        url += '?';
        var cri = objParam.criteria || {};
        var urlParams = [];
        for (var cr in cri) {
            if (cri.hasOwnProperty(cr) == false) continue;
            urlParams.push(cr + '=' + cri[cr]);
        }
        url += urlParams.join('&');
    }

    fnName = fnName || 'sendGet';
    this.send[fnName]({
        isParserResult: objParam.isParserResult,
        call: objParam.call,
        url: url
    });
};

/*上传*/
request.prototype.upload = function (objParam) {
    var criteria = objParam.criteria || {};
    var fileParamName = 'file';
    var realCriteria = {}
    realCriteria[fileParamName] =
        this.fs.readFileSync(criteria.path);
    this.send.upload({
        criteria: realCriteria,
        call: objParam.call,
        url: objParam.url
    });
};

/*下载*/
request.prototype.down = function (objParam) {
    var criteria = objParam.criteria || {};
    var key = criteria.key;
    var fileName = key.substr(30);
    objParam.fileName = fileName;
    this.send.down(objParam);
};

/*根据参数进行下载  objParam的释义同sendPost*/
request.prototype.pdownloadByParam = function (objParam) {
    var requestParam = objParam.criteria;
    this.send.pdownloadByParam({
        criteria: requestParam,
        call: objParam.call,
        url: objParam.url
    });
};


/*对请求参数进行处理*/
request.prototype.constructorParam = function (criteria) {
    return criteria;
};

request.prototype.constructorCriteriaStr = function (criteria) {
    var cri = criteria || {};
    //var cri = criteria instanceof Array == true ? criteria : [criteria];
    return encodeURIComponent(JSON.stringify(cri));
};





/*----------------------------发送请求并解析返回----------------------------*/
function send() {
    this.request = require('../public_nodejs/request');
    this.http = require('http');
};

/*
*普通post请求
*参数objParam    object类型，包括属性如下：
*   isParserResult   是否解析结果
*   criteria(非必须) object类型
*   call(非必须)  function类型
*   url 请求地址
*示例sendPost({
    criteria:{name:'111'},call:function(err,result){},fn:'getUser'
})
*/
send.prototype.sendPost = function (objParam) {
    var _this = this;
    _this.request.post({
        url: objParam.url,
        headers: {
            "content-type": "application/json",
        },
        body: JSON.stringify(objParam.criteria)
    }, function (err, httpResponse, body) {
        var obj = _this.parseResponse(err, body, httpResponse, objParam.isParserResult);
        if (typeof objParam.call === 'function')
            objParam.call(obj.err, obj.data, obj.count);
    });
};


/*get请求  参数释义同sendPost*/
send.prototype.sendGet = function (objParam) {
    var _this = this;
    _this.request.get({
        url: objParam.url
    }, function (err, httpResponse, body) {
        var obj = _this.parseResponse(err, body, httpResponse, objParam.isParserResult);
        if (typeof objParam.call === 'function')
            objParam.call(obj.err, obj.data);
    });
};

/*上传 formData里的file为数组，数组内多个文件流时即批量上传 */
send.prototype.upload = function (objParam) {
    var _this = this;
    _this.request({
        har: {
            url: objParam.url,
            method: 'POST',
            headers: [{
                name: 'content-type',
                value: 'application/octet-stream'
            }],
            postData: {
                mimeType: 'application/octet-stream',
                text: objParam.criteria.file
            }
        }
    }, function (err, httpResponse, body) {
        var obj = err ? {
            err: err
        } : _this.parseResponse(err, body, httpResponse);
        if (typeof objParam.call == 'function') objParam.call(obj.err, obj.data);
    });
};

/*下载*/
send.prototype.down = function (objParam) {
    var _this = this;
    var fileName = objParam.fileName || '_temp';
    var bufArr = [];
    var isErr = false;
    var res;
    var req = this.request.get({
        url: objParam.url,
        headers: {
            "content-type": "application/json",
            'cache-control': 'no-cache'
        }
    });
    req.on('error', function (e) {
        isErr = true;
        objParam.call('下载出错: ' + e.message);
    }).on('response', function (response) {
        res = response;
        if (res.statusCode !== 200 && res.statusCode !== 201) {
            isErr = true;
            return objParam.call('下载时连接错误，状态码：' + res.statusCode);
        }
    }).on('data', function (data) {
        bufArr.push(data);
    }).on('end', function () {
        if (isErr) return;
        objParam.call(null, {
            fileName: fileName,
            buffer: bufArr
        });
    });
    //req.end();
};

/*根据参数进行下载  objParam的释义同sendPost*/
send.prototype.pdownloadByParam = function (objParam) {
    var bufArr = [];
    var isErr = false;
    var res;
    var req = this.request.post({
        url: objParam.url,
        headers: {
            "content-type": "application/json",
            'cache-control': 'no-cache'
        },
        body: JSON.stringify(objParam.criteria)
    }).on('error', function (e) {
        isErr = true;
        objParam.call('下载出错: ' + e.message);
    }).on('response', function (response) {
        res = response;
        if (res.statusCode !== 200 && res.statusCode !== 201) {
            isErr = true;
            return objParam.call('下载时连接错误，状态码：' + res.statusCode);
        }
    }).on('data', function (data) {
        bufArr.push(data);
    }).on('end', function () {
        if (isErr) return;
        var fileName = '';
        var dispositionArr = (res.headers['content-disposition'] || '').split(';');
        for (var i = 0; i < dispositionArr.length; i++) {
            var currDis = dispositionArr[i];
            if (currDis.indexOf('filename') > -1) {
                fileName = currDis;
                break;
            }
        }

        fileName = fileName.split('=')[1] || '';
        fileName = decodeURI(fileName);
        fileName = fileName.replace(/\"/g, '') || '_temp';
        objParam.call(null, {
            fileName: fileName,
            buffer: bufArr
        });
    });
};

send.prototype.parseResponse = function (responseErr, responseTxt, _httpInfo, isParse) {
    if (responseErr) return {
        err: responseErr
    };
    if (_httpInfo.statusCode != 200) return {
        err: responseTxt || '后台返回异常'
    };
    isParse = isParse === false ? false : true;
    if (!isParse) return {
        data: responseTxt
    };
    var responseObj = {};
    try {
        responseObj = (typeof responseTxt == 'object' ? responseTxt : JSON.parse(responseTxt)) || {};
    } catch (e) {
        responseObj = {};
    } finally {
        var err = null,
            data = null,
            errStr = null;
        if (responseObj.Result !== 'success') errStr = responseObj.ResultMsg || responseObj.Result || '后台返回异常';
        var content = responseObj.Content || [responseObj.Item];
        var returnObj = {
            err: errStr,
            data: content
        };
        if (responseObj.Count || responseObj.Count === 0)
            returnObj.count = responseObj.Count;
        return returnObj;
    }
};
