// common/log4logger.js
var log4 = function () {
    this.log4js = require('../public_nodejs/log4js');
    this.fs = require('fs');
};

log4.prototype.CreateLog = function (app, appPath) {

    appPath = appPath.replace(/\\/g, '/');
    var logPath = appPath + '/logs';
    var isExists = this.fs.existsSync(logPath);
    if (isExists == false)
        this.fs.mkdirSync(logPath);
    var fileName = logPath + '/log.log';
    this.log4js.configure({
        appenders: [
            { type: 'console' },
           {
               "type": "dateFile",      //dateFile的意思为每天会创建一个新文件
               'category': 'log_file',
               "filename": fileName,
               "pattern": "yyyy-MM-dd.log",
               "alwaysIncludePattern": false
           }
        ],
        replaceConsole: true,
        levels: {
            log_file: 'WARN'        //日志级别 ALL TRACE DEBUG INFO WARN ERROR FATAL OFF
        }
    });
    /*logger具有八个方法：all、trace、debug、info、warn、error、fatal、off  其中warn、error才会往日志文件了写入信息*/
    var logger = this.log4js.getLogger('log_file');
    global.pLogger = logger;
    console.error = logger.error.bind(logger);
    console.warn = logger.warn.bind(logger);
    if (app)
        app.use(this.log4js.connectLogger(logger, { format: '\r\n\r\n:method :url\r\n\r\n\r\n\r\n' }));
};

module.exports = new log4();
